File: dataset_reduzido/avatar/Python/Code/atcoder_AGC011_C.py
import queue
n , m = map ( int , input ( ).split ( ) )
vis , ci , cb , cc = [ 0 ] * ( n + 1 ) , 0 , 0 , 0
g = [ [ ] for i in range ( n + 1 ) ]
def dfs ( x ) :
    stk , flag = queue.LifoQueue ( ) , True
    stk.put ( ( x , 1 ) )
    while not stk.empty ( ) :
        u , col = stk.get ( )
        if vis [ u ] :
            flag &= ( vis [ u ] == col )
            continue
        vis [ u ] = col
        for i in g [ u ] :
            stk.put ( ( i , 3 - col ) )
    return flag
for i in range ( m ) :
    u , v = map ( int , input ( ).split ( ) )
    g [ u ] += [ v ]
    g [ v ] += [ u ]
for i in range ( 1 , n + 1 ) :
    if vis [ i ] == 0 :
        if len ( g [ i ] ) == 0 :
            ci += 1
        else :
            if dfs ( i ) :
                cb += 1
            else :
                cc += 1
print ( ci * ci + 2 * ci * ( n - ci ) + cc * cc + 2 * cb * cc + 2 * cb * cb )

File: output/LLaMa/avatar/Python/Java/atcoder_AGC011_C.java
import java.io.*;
import java.util.*;

public class Main {
    static ArrayList<ArrayList<Integer>> g = new ArrayList<>();
    static int n, m;
    static int[] vis = new int[10001];
    static int ci, cb, cc;

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        n = scanner.nextInt();
        m = scanner.nextInt();
        for (int i = 0; i <= n; i++) {
            g.add(new ArrayList<>());
        }
        for (int i = 1; i <= n; i++) {
            vis[i] = 0;
        }
        for (int i = 0; i < m; i++) {
            int u = scanner.nextInt();
            int v = scanner.nextInt();
            g.get(u).add(v);
            g.get(v).add(u);
        }
        for (int i = 1; i <= n; i++) {
            if (vis[i] == 0) {
                if (g.get(i).isEmpty()) {
                    ci++;
                } else {
                    if (dfs(i)) {
                        cb++;
                    } else {
                        cc++;
                    }
                }
            }
        }
        System.out.println(ci * ci + 2 * ci * (n - ci) + cc * cc + 2 * cb * cc + 2 * cb * cb);
    }

    public static boolean dfs(int x) {
        Stack<int[]> stk = new Stack<>();
        stk.push(new int[]{x, 1});
        boolean flag = true;
        while (!stk.isEmpty()) {
            int[] temp = stk.pop();
            int u = temp[0];
            int col = temp[1];
            if (vis[u] != 0 && vis[u] != col) {
                flag = false;
            } else if (vis[u] != 0) {
                continue;
            }
            vis[u] = col;
            for (int i : g.get(u)) {
                stk.push(new int[]{i, 3 - col});
            }
        }
        return flag;
    }
}